csrc/optimization/layout_optimization.cpp

```diff
@@ -0,0 +1,45 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/layout_inference.h>
+#include <optimization/layout_optimization.h>
+
+namespace nvfuser::optimization {
+
+void LayoutOptimizationPass::runPass(Fusion* fusion) {
+  std::unordered_map<const TensorView*, MemoryFormat> stride_mapping =
+      inferenceMemoryFormat(fusion);
+
+  for (auto out_val : fusion->outputs()) {
```

Comment:
Why are we only setting the allocation domain of the output of the unsegmented fusion? If my fusion is:
```C++
input: t0
output: t2
t1 = segmenter_set(t0)
t2 = segmenter_set(t1)
```
and `t0` is NHWC, then `t2` will be set as NHWC correctly, but `t1` will remain NCHW? Then both segments will use transpose scheduler instead of pointwise scheduler? This behavior doesn't make sense to me. Should we just set all tensor's allocation domain here, and inside the `scheduleXXX` (where `XXX` = `Pointwise`, `Transpose`, etc.) functions, reset the intermediate tensor's allocation before doing any real schedule?

Reply:
That's a very reasonable ask. i.e. not just for `segmenter_set`, we might as well want to support general segmented fusion.

But it's currently not in the scope for initial PRs. I'm leaning towards that we'd want to push this pass to each scheduler to handle, which we'll naturally be able to handle segment boundaries.

Reply:
Sure, it don't have to be done in this PR. But why do you think
```
we'd want to push this pass to each scheduler
```
? Schedulers needs to know the stride order to decide if it can schedule the fusion or not.

Reply:
> ? Schedulers needs to know the stride order to decide if it can schedule the fusion or not.

Sure, if we do want a certain tensor to return in a specific stride order, that needs to be set prior to passing things to scheduler.

But for tensor with an empty allocation domain, it felt natural for a scheduler to come in and return something like. "i can schedule this fusion, meanwhile, I'm going to fill in empty allocation domain to something I can generate efficient kernel for"

Reply:
> I'm going to fill in empty allocation domain

Isn't this information ambiguous? For example, for a segment `T2 = set(T1)` where neither the allocation domain of `T1` and `T2` is set, the pointwise scheduler can fill either NHWC or NCHW, or WNHC, etc., it is hard to infer which one is preferred without the global stride order information from propagation.

Reply:
>  the global stride order information from propagation.

I'm not saying that the decision should be made merely within a `segment`. You should still see information propagated from your producer segment.

i.e. `T2 = set(T1)` would have a producer segment that hopefully will try to mark allocation domain on T1. So within each segment we are propagating only from inputs to outputs and that shouldn't leave any ambiguity in the rule. I think this works better with smaller segments using a small set of propagation rules.

I think what you were proposing is more aligned with @jacobhinkle's suggestion, an overall pass the does both forward and backward propagation. Which I'm very open minded as a future refactor.